libs/core/src/main/java/org/elasticsearch/core/internal/provider/ProviderLocator.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.core.internal.provider;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.security.AccessController;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+import java.util.Objects;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.function.Supplier;
+
+/**
+ * A provider locator that finds the implementation of the specified provider.
+ *
+ * <p> A provider locator is given a small recipe, in the form of constructor arguments, which it uses to find the required provider
+ * implementation.
+ *
+ * <p> When run as a module, the locator will load the provider implementation as a module, in its own module layer.
+ * Otherwise, the provider implementation will be loaded as a non-module.
+ *
+ * @param <T> the provider type
+ */
+public final class ProviderLocator<T> implements Supplier<T> {
+
+    private final String providerName;
+    private final Class<T> providerType;
+    private final String providerModuleName;
+    private final Set<String> missingModules;
```

Comment:
I assume this is held over from the module branch, but not applicable to master right now?

Reply:
Yes, I'll remove the modules stuff from this PR.